### PR TITLE
Move heap into its own internal package

### LINKED ISSUE #71863: [scheduler cleanup phase 2]: Rename `pkg/scheduler/util` to `pkg/scheduler/internal/podinfo`
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.

Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

If you contributed in phase 1, please give others a chance before claiming this issue. You can also help out by reviewing the phase 2 PRs.

This package contains only functions related to pods. Having “util” as a name is generally discouraged, since it doesn’t tell readers anything about what is in that package. This change will also mirror the change to create `pkg/scheduler/nodeinfo` which will have a bunch of node-related types and functions.

`GetPodPriority` from this package is used by `pkg/kubelet/eviction`. In order to make our package internal, we should do one of the following. (1 is my preference)

1. Copy the function into kubelet and remove the dependency
2. Move the function into a different area like `apimachinery` or `client-go`
3. Make the funcion an official part of the scheduler’s public go libraries

/sig scheduling
/kind cleanup
/help
/good-first-issue

### PR DESCRIPTION
**Pull request by hprateek43**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR is in continuation of PR #76990. This PR aims to move heap into its internal package as part of #71863 
**Which issue(s) this PR fixes**:

Fixes #71863 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by hprateek43** [comment]
@alculquicondor Can you give an `ok-to-test` ?

**Review by draveness — COMMENTED** [review]
/ok-to-test
/cc @alculquicondor

File: pkg/scheduler/internal/heap/heap.go [hunk 1]
```diff
@@ -256,3 +256,7 @@ func NewHeapWithRecorder(keyFn KeyFunc, lessFn LessFunc, metricRecorder metrics.
 		metricRecorder: metricRecorder,
 	}
 }
+
```
Comment (alculquicondor) [inline pkg/scheduler/internal/heap/heap.go]: `NewHeap` should just be `New`. `heapData` should just be `data`

File: pkg/scheduler/internal/heap/heap.go [hunk 2]
```diff
@@ -256,3 +256,7 @@ func NewHeapWithRecorder(keyFn KeyFunc, lessFn LessFunc, metricRecorder metrics.
...
+// LessFunc is a function that receives two items and returns true if the first
+// item should be placed before the second one when the list is sorted.
+type LessFunc func(item1, item2 interface{}) bool
```
Comment (alculquicondor) [inline pkg/scheduler/internal/heap/heap.go]: you can make this a private alias instead. There's not much benefit on having it as defined type.

File: pkg/scheduler/util/utils.go [hunk 3]
```diff
@@ -91,13 +92,9 @@ func GetEarliestPodStartTime(victims *api.Victims) *metav1.Time {
 // SortableList is a list that implements sort.Interface.
 type SortableList struct {
 	Items    []interface{}
-	CompFunc LessFunc
+	CompFunc heap.LessFunc
```
Comment (alculquicondor) [inline pkg/scheduler/util/utils.go]: nope :)
Either `LessFunc` comes from a common package, or each package defines their own. A private alias should be enough.

**Comment by hprateek43** [comment]
@alculquicondor Can you review the changes?

File: pkg/scheduler/util/utils.go (hunk 3 above)
Comment (hprateek43) [reply pkg/scheduler/util/utils.go]: Changes as per recommendation

File: pkg/scheduler/internal/heap/heap.go (hunk 2 above)
Comment (hprateek43) [reply pkg/scheduler/internal/heap/heap.go]: Chnages as per recommendation.

Comment (alculquicondor) [reply pkg/scheduler/internal/heap/heap.go]: Did you try making it a **private** alias instead?

**Comment by ahg-g** [comment]
> **What type of PR is this?**
> 
> > Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
> > /kind api-change
> > /kind bug
> > /kind cleanup
> > /kind design
> > /kind documentation
> > /kind failing-test
> > /kind feature
> > /kind flake
> 
> **What this PR does / why we need it**:
> This PR is in continuation of PR #76990. This PR aims to move heap into its internal package as part of #71863
> **Which issue(s) this PR fixes**:
> 
> Fixes #71863

Does this PR actually fix the issue, or is it just part of it?

**Comment by hprateek43** [comment]
> > **What type of PR is this?**
> > > Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
> > > /kind api-change
> > > /kind bug
> > > /kind cleanup
> > > /kind design
> > > /kind documentation
> > > /kind failing-test
> > > /kind feature
> > > /kind flake
> > 
> > 
> > **What this PR does / why we need it**:
> > This PR is in continuation of PR #76990. This PR aims to move heap into its internal package as part of #71863
> > **Which issue(s) this PR fixes**:
> > Fixes #71863
> 
> Does this PR actually fix the issue, or is it just part of it?

This PR solves a part of the issue. The previous PR was split into 3 PRs

**Comment by ahg-g** [comment]
/approve 
/hold 

for @alculquicondor to lgtm

**Comment by ahg-g** [comment]
> > > **What type of PR is this?**
> > > > Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
> > > > /kind api-change
> > > > /kind bug
> > > > /kind cleanup
> > > > /kind design
> > > > /kind documentation
> > > > /kind failing-test
> > > > /kind feature
> > > > /kind flake
> > > 
> > > 
> > > **What this PR does / why we need it**:
> > > This PR is in continuation of PR #76990. This PR aims to move heap into its internal package as part of #71863
> > > **Which issue(s) this PR fixes**:
> > > Fixes #71863
> > 
> > 
> > Does this PR actually fix the issue, or is it just part of it?
> 
> This PR solves a part of the issue. The previous PR was split into 3 PRs. I think I should remove the fixes part from the PR description

please update the first comment and replace "Fixes #71863" with "Part of #71863", otherwise this will cause the issue to close.

**Comment by alculquicondor** [comment]
You can squash now

/lgtm

**Comment by alculquicondor** [comment]
I think the bots are back online

/unhold

**Pull request merged**